### PR TITLE
AWS EKS example workflow to deploy NIM for LLMs

### PR DESCRIPTION
**Pull request by dpadmanabhan03**
- This PR introduces an AWS EKS example workflow to deploy NIM for LLMs.
- The implementation includes AWS high level infrastructure diagram, vpc stack, eks cluster stack and efs stack.
- The workflow leverages existing helm charts to deploy llama 3 8B NIM.
- It also includes prometheus monitoring configuration setup and an ingress controller configuration.

### CODE REVIEW

File: cloud-service-providers/aws/eks/README.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
+# NIM-EKS: NVIDIA Inference Microservices on AWS EKS
+This GitLab repository is dedicated to testing NIM (NVIDIA Inference Microservices) on AWS EKS (Elastic Kubernetes Service).
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/README.md:2]: Remove word "GitLab"

Comment (supertetelman) [inline cloud-service-providers/aws/eks/README.md:2]: Just refer to NIM as "NVIDIA NIM", do not spell it out.

File: cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+  constructor(scope: Construct, id: string, props: EksClusterStackProps) {
+    super(scope, id, props);
+
+    // Define IAM policy statement to allow list access to eks cluster
+    const eksPolicyStatement = new iam.PolicyStatement({
+      effect: iam.Effect.ALLOW,
+      actions: ["eks:*"],
+      resources: ["*"],
+    });
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts:28]: I'm not as familiar with TypeScript style guides, but it looks like there should be a newline in a few places in these two typescript files for better readability.

File: cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    // Define IAM policy statement to describe cloudformatiom stacks
+    const cfnPolicyStatement = new iam.PolicyStatement({
+      effect: iam.Effect.ALLOW,
+      actions: ["cloudformation:DescribeStacks"],
+      resources: ["*"],
+    });
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts:34]: And here?

File: cloud-service-providers/aws/eks/README.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+2: Refer this to prepare your local environment to run cdk
+https://docs.aws.amazon.com/cdk/v2/guide/getting_started.html
+
+There are four stacks in the nim-eks-cdk/lib directory. You can either deploy all of them using `cdk deploy --all` or deploy each stack by specifying the stack name
+like `cdk deploy vpc-stack`
+
+Stack Deployment Order:
+
+    cdk deploy vpc-stack
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/README.md:17]: Put these in a code block if they are meant to be run.

```
<code>
```

File: cloud-service-providers/aws/eks/README.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+ClusterAdmin user is created by the cdk. Create access keys for this user to run kubectl and helm commands. Once secret credentials are created run the following command to manage the cluster
+
+    aws eks update-kubeconfig --name <eks-cluster-name> --region <region-code> --profile <profile-name-if-any>
+
+Note:  Ensure that you are in the cloud-service-providers/aws/eks directory for the rest of this guide.
+
+## Cluster setup for inference:
+
+1: Install NVIDIA Device Plugin: Install NVIDIA device plugins to run GPU workloads. Check CUDA base image version for compatibility.
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/README.md:31]: Is there a reason we are directly installing the device plugin and not the GPU Operator?

File: cloud-service-providers/aws/eks/README.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    export NGC_API_KEY=<NGC_API_KEY>
+    kubectl create secret generic ngc-api --from-literal=NGC_CLI_API_KEY=$NGC_API_KEY
+
+3: Create Docker Registry Secret:
+
+    kubectl create secret generic registry-secret  \
+    --from-file=.dockerconfigjson=/path/to/.docker/config.json \
+    --type=kubernetes.io/dockerconfigjson
+
+Note: If you are not logged in to the docker nvcr.io registry run docker login nvcr.io using the NGC API key
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/README.md:54]: We should refer to nvcr.io as a "container registry", not a "docker registry"

File: cloud-service-providers/aws/eks/perf/gen-ai-perf.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  name: triton
+  labels:
+    app: triton
+spec:
+  containers:
+  - name: triton
+    image: nvcr.io/nvidia/tritonserver:24.04-py3-sdk
+    resources:
+      limits:
+        nvidia.com/gpu: 1
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/perf/gen-ai-perf.yaml:13]: If this is just hitting the running NIM, do we need to give the Triton pod a GPU resource?

**Review by supertetelman — CHANGES_REQUESTED** [review]
Please make the suggested changes in the review and we can merge through. Mostly a few minor style/branding issues.

File: cloud-service-providers/aws/eks/README.md (hunk 5 above)
Comment (dpadmanabhan03) [reply cloud-service-providers/aws/eks/README.md:31]: This example uses Amazon EKS optimized Amazon Linux AMIs (Amazon Linux 2 AMI). It's default configuration isn't compatible with GPU operator. https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/amazon-eks.html

File: cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts (hunk 2 above)
Comment (dpadmanabhan03) [reply cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts:23]: Added new line.

File: cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts (hunk 3 above)
Comment (dpadmanabhan03) [reply cloud-service-providers/aws/eks/nim-eks-cdk/lib/eks-cluster-stack.ts:30]: Updated.

**Comment by hustshawn** [comment]
Hi @dpadmanabhan03 , it's glad to see you are working on the same as me on this https://github.com/awslabs/data-on-eks/issues/560 , my implementation will be Terraform based in our Data on EKS patterns. 

I just raised an issue (https://github.com/NVIDIA/nim-deploy/issues/24) in this repo, see if you facing the same? cuz both you and me implement on top of EKS. I am looking forward to the insights from NVIDIA side.

File: cloud-service-providers/aws/eks/README.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
+# NIM-EKS: NVIDIA Inference Microservices on AWS EKS
```
Comment (supertetelman) [inline cloud-service-providers/aws/eks/README.md:1]: This should just read "NVIDIA NIM on AWS EKS", no need to spell NIM out.

**Review by supertetelman — CHANGES_REQUESTED** [review]
One more minor branding change, then we can merge this.

File: cloud-service-providers/aws/eks/README.md (hunk 8 above)
Comment (dpadmanabhan03) [reply cloud-service-providers/aws/eks/README.md]: Updated.

**Review by supertetelman — APPROVED** [review]
Great guide!

**Pull request merged**